core-primitives/stf-executor/src/executor_tests.rs

```diff
@@ -0,0 +1,317 @@
+	let (stf_executor, _, state_handler) = stf_executor();
+	let key = "my_key".encode();
+	let value = "my_value".encode();
+	let old_state_hash = state_hash(&state_handler.load_initialized(&shard).unwrap());
+
+	// when
+	let batch_execution_result = stf_executor
+		.propose_state_update(
+			&vec![],
+			&ParentchainHeaderBuilder::default().build(),
+			&shard,
+			Duration::ZERO,
+			|mut state| {
+				state.insert(key.clone(), value.clone());
+				state
+			},
+		)
+		.unwrap();
+
+	// then
+	assert_eq!(old_state_hash, batch_execution_result.state_hash_before_execution);
+
+	// Ensure that state has been updated.
+	let old_state = state_handler.load_initialized(&shard).unwrap();
+	let retrieved_value = batch_execution_result.state_after_execution.get(key.as_slice()).unwrap();
+	assert_eq!(*retrieved_value, value);
+	// Ensure that state has not been actually written.
+	assert_ne!(old_state, batch_execution_result.state_after_execution);
+}
+
+pub fn execute_timed_getters_batch_executes_if_enough_time() {
+	// given
+	let sender = endowed_account();
+	let trusted_getter = TrustedGetter::free_balance(sender.public().into()).sign(&sender.into());
+	let (stf_executor, _, state_handler) = stf_executor();
+	let (_, shard) = init_state_and_shard_with_state_handler(state_handler.as_ref());
+	let execution_duration = Duration::from_secs(10000);
+
+	// when
+	assert!(panic::catch_unwind(|| {
+		stf_executor.execute_timed_getters_batch(
+			&vec![trusted_getter.clone()],
+			&shard,
+			execution_duration,
+			|trusted_getter_signed: &TrustedGetterSigned,
+			 _state_result: Result<Option<Vec<u8>>>| {
+				// then
+				assert_eq!(*trusted_getter_signed, trusted_getter);
+				panic!("test should enter here");
+			},
+		)
+	})
+	.is_err());
+}
+
+pub fn execute_timed_getters_does_not_execute_more_than_once_if_not_enough_time() {
+	// given
+	let sender = endowed_account();
+	let trusted_getter =
+		TrustedGetter::free_balance(sender.public().into()).sign(&sender.clone().into());
+	let trusted_getter_two =
+		TrustedGetter::reserved_balance(sender.public().into()).sign(&sender.into());
+	let (stf_executor, _, state_handler) = stf_executor();
+	let (_, shard) = init_state_and_shard_with_state_handler(state_handler.as_ref());
+	let execution_duration = Duration::ZERO;
+
+	// when
+	stf_executor
+		.execute_timed_getters_batch(
+			&vec![trusted_getter.clone(), trusted_getter_two],
+			&shard,
+			execution_duration,
+			|trusted_getter_signed: &TrustedGetterSigned,
+			 _state_result: Result<Option<Vec<u8>>>| {
+				// then (second getter should not be executed)
+				assert_eq!(*trusted_getter_signed, trusted_getter);
+			},
+		)
+		.unwrap();
+}
+
+pub fn execute_timed_getters_batch_returns_early_when_no_getter() {
+	// given
+	let shard = ShardIdentifier::default();
+	let (stf_executor, _, _) = stf_executor();
+	let execution_duration = Duration::from_secs(10000);
+
+	// when
+	stf_executor
+		.execute_timed_getters_batch(
+			&vec![],
```

Comment:
I want to hear your opinions on this: Do you prefer using `vec![]`, `Vec::new()`, `Vec::default()` or `Default::default()` for the empty vector?

Reply:
Hm. Good question. I'd not opt for `default()` a default does not necessarily mean an empty vector and is longer in terms of spelling than the first two options.
I'd probably opt for `Vec::new()`. It turned out as `vec![]` here because I wrote a non-empty vector first. Removing something is faster than rewriting 😄 

Do you think we should stay consistent? I don't have a strong opinion here.

Reply:
I don't have a strong opinion either - but I think we should agree on 1 or 2 options, in order to improve readability. So maybe `vec![]` and `Vec::new()` but not `default()`?

Reply:
Agreed - expect for actual default value construction. There, I think it makes sense to construct a default vector as well.